### PR TITLE
Exclude Directories

### PR DESCRIPTION
**Pull request by mmattel**
Feature PR, references: Exclude Directories (merged, for professional…… Storage use) in owncloud/core/pull/19235

Adds the ability to define directories in config.php which will not be further processed (scanned, shown, created ect). These directories exist in the source filesystem but will not handled in nextcloud.

Usecase: eg. filesystems which have the ability of snapshots have a defined folder naming for these snapshots. Processing them is in most cases not sensful.

For a detailed description pls see also the documentation PR.

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/Directory.php [hunk 1]
```diff
@@ -107,7 +107,17 @@ public function __construct($view, $info, $tree = null, $shareManager = null) {
 	 */
 	public function createFile($name, $data = null) {
 
+		# the check here is necessary, because createFile uses put covered in sabre/file.php
```
Comment (nickvergessen) [inline apps/dav/lib/Connector/Sabre/Directory.php]: `#` comments are deprecated in php 🙊

File: apps/dav/lib/Connector/Sabre/Directory.php [hunk 2]
```diff
@@ -107,7 +107,17 @@ public function __construct($view, $info, $tree = null, $shareManager = null) {
...
+		# the check here is necessary, because createFile uses put covered in sabre/file.php 
+		# and not touch covered in files/view.php
+		if (\OC\Files\Filesystem::isForbiddenFileOrDir($name)) {
+			throw new \Sabre\DAV\Exception\Forbidden();
+		}
 		try {
+			# the check here is necessary, because createFile uses put covered in sabre/file.php 
+			# and not touch covered in files/view.php
+			if (\OC\Files\Filesystem::isForbiddenFileOrDir($name)) {
+				throw new \Sabre\DAV\Exception\Forbidden();
```
Comment (nickvergessen) [inline apps/dav/lib/Connector/Sabre/Directory.php]: either this or the call above is wrong

**Review by LukasReschke — COMMENTED** [review]
I think that should be handled on storage system level and not a global setting. I'm not really in favour of that as it is quite an unexpected behaviour for users and likely won't be tested properly since it's a setting for edge cases.

If that would be handled at least on the storage level layer than that would at least only apply to the minimum storages.

Also what's the use-case behind this? What is the hardware in case exactly? Can't it be configured in another way?

cc @icewind1991

File: apps/dav/lib/Connector/Sabre/ObjectTree.php [hunk 3]
```diff
@@ -109,13 +109,19 @@ public function cacheNode(Node $node) {
...
 		if (!$this->fileView) {
 			throw new \Sabre\DAV\Exception\ServiceUnavailable('filesystem not setup');
 		}
 
+		// check the path, also called when the path has been entered manually eg via a file explorer
+		if (\OC\Files\Filesystem::isForbiddenFileOrDir($path)) {
+			throw new \Sabre\DAV\Exception\Forbidden();
+		}
+		
 		$path = trim($path, '/');
```
Comment (nickvergessen) [inline apps/dav/lib/Connector/Sabre/ObjectTree.php]: Swap the order of this assignment and the check before, then you can get rid of the second check below

File: apps/dav/lib/Connector/Sabre/ObjectTree.php [hunk 4]
```diff
@@ -130,6 +136,11 @@ public function getNodeForPath($path) {
 			}
 		}
 
+		// check the path, also called when the path has been entered manually eg via a file explorer
+		if (\OC\Files\Filesystem::isForbiddenFileOrDir($path)) {
```
Comment (nickvergessen) [inline apps/dav/lib/Connector/Sabre/ObjectTree.php]: drop this one

File: lib/private/Files/Filesystem.php [hunk 5]
```diff
@@ -565,29 +565,26 @@ static public function isValidPath($path) {
 	 *
 	 * @param array $data from hook
 	 */
-	static public function isBlacklisted($data) {
```
Comment (nickvergessen) [inline lib/private/Files/Filesystem.php]: this method might be used by other apps?

File: lib/private/Files/Filesystem.php [hunk 6]
```diff
@@ -604,6 +601,55 @@ static public function isIgnoredDir($dir) {
...
+		$blacklist = array_map('strtolower', $blacklist);
+		$match = array_intersect($path_parts, $blacklist);
+		if ($match) {
+			return true;
+		}
+		return false;
+	}
+
+	/**
 	/**
```
Comment (nickvergessen) [inline lib/private/Files/Filesystem.php]: duplication === error

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Comment by mmattel** [comment]
@LukasReschke 
- There was a long discussion upfront on this (~2 years) and finally it was agreed to implement/merge this on ownCloud.
- There are many professional Storage vendors like NetApp, EMC, HDS ect but also public available filesystem like ZFS, BTRFS who are capable of snapshots. Just read the documentation linked. If you have further questions, let me know.
- No that can´t be done on a storage level because you do not know which storage you will mount and from which provider it will come. In reality you always mount a filesystem which provides folders and files. And snapshots are special folders with a defined but different naming by each vendor.

> ...quite an unexpected behaviour for users...

- It is just the other way round. Snapshots are mostly used for backup/restore purposes. Those folders containing the snapshots have in most cases no value for the enduser but the admin. Please just see also the math behind in the documentation referenced.
- From a company / professional use point of view I can tell you that excluding directories is key for owncloud/nextcloud.

File: apps/dav/lib/Connector/Sabre/Directory.php (hunk 1 above)
Comment (mmattel) [inline apps/dav/lib/Connector/Sabre/Directory.php]: corrected

File: apps/dav/lib/Connector/Sabre/Directory.php (hunk 2 above)
Comment (mmattel) [inline apps/dav/lib/Connector/Sabre/Directory.php]: corrected

File: apps/dav/lib/Connector/Sabre/ObjectTree.php (hunk 3 above)
Comment (mmattel) [inline apps/dav/lib/Connector/Sabre/ObjectTree.php]: corrected

File: apps/dav/lib/Connector/Sabre/ObjectTree.php (hunk 4 above)
Comment (mmattel) [inline apps/dav/lib/Connector/Sabre/ObjectTree.php]: corrected

File: lib/private/Files/Filesystem.php (hunk 5 above)
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: Yes, when I originally created the PR in owncloud, I have found one app (but I do not remember anymore...) that was using this call. Therefore I left it and forwarded the call to the new function.

File: lib/private/Files/Filesystem.php (hunk 6 above)
Comment (mmattel) [inline lib/private/Files/Filesystem.php]: corrected

**Comment by LukasReschke** [comment]
@icewind1991 Please comment on the storage wrapper approach. THX.

**Comment by rullzer** [comment]
I agree. This solution is way to hacky for my taste. And it is not something that core should take care of. A simple POC to do this with an app in https://github.com/rullzer/files_excludedirs

**Comment by LukasReschke** [comment]
Closing as it should be done via an app as the PoC linked from https://github.com/nextcloud/server/pull/1627#issuecomment-251776426 shows.

**Pull request closed without merging**